src/rtc_pl031.rs

```diff
@@ -0,0 +1,639 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! ARM PL031 Real Time Clock
+//!
+//! This module implements a PL031 Real Time Clock (RTC) that provides a long
+//! time base counter. This is achieved by generating an interrupt signal after
+//! counting for a programmed number of cycles of a real-time clock input.
+//!
+use std::result::Result;
+use std::time::Instant;
+
+use crate::Trigger;
+
+// As you can see in
+//  https://static.docs.arm.com/ddi0224/c/
+//  real_time_clock_pl031_r1p3_technical_reference_manual_DDI0224C.pdf
+//  at section 3.2 Summary of RTC registers, the total size occupied by this
+//  device is 0x000 -> 0xFFC + 4 = 0x1000.
+
+// From 0x0 to 0x1C we have following registers:
+const RTCDR: u64 = 0x0; // Data Register.
```

Comment:
Why do we define the registers as u64? It looks like we are then using them as u32, and we need to do a conversion which would be unsafe if the registers would not fit in u32.

Reply:
1. I've incorporated using the same register definitions as the spec.
2. I don't know why `u64`... I've changed them to `u32`.

Reply:
actually, these are register offsets, so they don't even need to be u32. they can just be the smallest size that can fit the values. As far as I can see, a `u16` works for all cases. @andreeaflorescu

Reply:
I find using u32 to be better in this particular case because we are expecting `data` to be `u32` ( 4 * u8).Picking a type where we need less types conversions means less possibilities for failures.

Reply:
@alindima and I synced offline. We don't need the offset to be a u32 because there is no operation on offsets that include other u32 values. We can thus make all offsets/registers definition u16. This would mean updates to all register definitions (and their examples), and we can also do it in a separate PR if that's too much @dodsonmg.

Reply:
I actually think this approach of converting an u64 to a smaller type is wrong. Let us think what happens in a scenario where a malicious guest actually sends us an u64. In Rust, as far as my knowledge, if you try to convert from a larger value into something that will not fit that value, **as** takes the lower bits of the number, disregarding the upper bits. With this approach we run the risk of either;
a) never knowing that we are actually dealing with a malicious guest or
b) logging an error containing a modified offset
With the initial approach (keeping the offsets the same type as what we are expected to receive from the guest) none of the 2 from above are possible. Additionally, if we decide we want to do the conversion,  what is recommended to use for doing these type of conversions is [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).

Reply:
Works for me.  Done.

Reply:
@dianpopa, in this case, it's up to the manager of the Device to handle these casts. In the case of Firecracker, the `BusDevice` trait constrains the offset to be u64, but for this particular device, there is no reason why we couldn't have the offsets to be a smaller type. This is also what the serial device does: https://github.com/rust-vmm/vm-superio/blob/19dc073cdaf3e12ab7c720c9d4d4abd34872a7f0/src/serial.rs#L365 

Even if we decide that the read/write functions should receive a u64, indeed we can just cast them using `try_from`

Reply:
> I actually think this approach of converting an u64 to a smaller type is wrong. Let us think what happens in a scenario where a malicious guest actually sends us an u64. In Rust, as far as my knowledge, if you try to convert from a larger value into something that will not fit that value, **as** takes the lower bits of the number, disregarding the upper bits. With this approach we run the risk of either;
> a) never knowing that we are actually dealing with a malicious guest or
> b) logging an error containing a modified offset
> With the initial approach (keeping the offsets the same type as what we are expected to receive from the guest) none of the 2 from above are possible. Additionally, if we decide we want to do the conversion, what is recommended to use for doing these type of conversions is [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).

Indeed. We should be checking that the guest is sending u16 as an offset. I don't think that this is a concern of the device emulation. The device emulation works with 16-bytes (as per the specification), so there is no need to use u64 here.

What do you think about handling the error case you are describing at the Bus level? The reason I believe this check to be better suited at the Bus level is because if the device sits on a PIO Bus, then there is no possibility for a malicious guest to send anything else then a `u16` offset because PIO only accepts `u16` values. This means that only if the device sits on a MMIO Bus the check will be needed.

This is not the case with RTC specifically as we only use it on aarch64, but generally speaking we should implement the emulation following the specs, and provide any adjustments to the types (i.e. converting from u64 to u16 and deal with errors in casting) at the level where the values are forwarded from the `Bus` to the emulation.

Reply:
> > I actually think this approach of converting an u64 to a smaller type is wrong. Let us think what happens in a scenario where a malicious guest actually sends us an u64. In Rust, as far as my knowledge, if you try to convert from a larger value into something that will not fit that value, **as** takes the lower bits of the number, disregarding the upper bits. With this approach we run the risk of either;
> > a) never knowing that we are actually dealing with a malicious guest or
> > b) logging an error containing a modified offset
> > With the initial approach (keeping the offsets the same type as what we are expected to receive from the guest) none of the 2 from above are possible. Additionally, if we decide we want to do the conversion, what is recommended to use for doing these type of conversions is [TryFrom](https://doc.rust-lang.org/std/convert/trait.TryFrom.html).
> 
> Indeed. We should be checking that the guest is sending u16 as an offset. I don't think that this is a concern of the device emulation. The device emulation works with 16-bytes (as per the specification), so there is no need to use u64 here.
> 
> What do you think about handling the error case you are describing at the Bus level? The reason I believe this check to be better suited at the Bus level is because if the device sits on a PIO Bus, then there is no possibility for a malicious guest to send anything else then a `u16` offset because PIO only accepts `u16` values. This means that only if the device sits on a MMIO Bus the check will be needed.
> 
> This is not the case with RTC specifically as we only use it on aarch64, but generally speaking we should implement the emulation following the specs, and provide any adjustments to the types (i.e. converting from u64 to u16 and deal with errors in casting) at the level where the values are forwarded from the `Bus` to the emulation.

Yup, handling the error at the bus level sounds just right 👍

Reply:
> Works for me. Done.

I responded to a stale comment here, so just to be explicit:  I've implemented the register offsets as `u16`.

It sounds like the conversation has agreed that this is correct for this device, and handling the translation from 64 bit addresses to a 16 bit register offset will be handled at the bus.